Authentication for the live app

AndrewDryga opened the pull request.

AndrewDryga committed:
Fix gitignore

AndrewDryga committed:
Add CSP and STS headers

AndrewDryga committed:
Update LiveView along with other deps

AndrewDryga committed:
tmp

AndrewDryga committed:
Do not allow to enable duplicates of auth providers

AndrewDryga committed:
Make vault a dependency for api/web containers

AndrewDryga committed:
Bump phoenix version

AndrewDryga committed:
Add actor name

AndrewDryga committed:
Do not allow to sign in if identity is deleted or actor is deleted or disabled

AndrewDryga committed:
Extend auth behavior to support IdP flows

AndrewDryga committed:
Bump Elixir/Erlang/Alpine versions

AndrewDryga committed:
Allow signing in using identity id along with provider identifier

AndrewDryga committed:
Fix elixir 1.5 compilation and warning

AndrewDryga committed:
Make all auth flows work and add controller tests

LiveView and acceptance tests are WIP

AndrewDryga committed:
Make sure that session is not transferrable between account paths

jamilbk commented on elixir/apps/web/lib/web/mailer/auth_email/sign_in_link.text.heex:
```diff
@@ -0,0 +1,9 @@
+please copy this link and open it in your browser.
+
+It is valid for 1 hour.
+If you didn't request for this, you can safely discard this email.
```
```suggestion
If you didn't request this, you can safely discard this email.
```

jamilbk commented on elixir/apps/web/lib/web/mailer/auth_email/sign_in_link.text.heex:
```diff
@@ -0,0 +1,9 @@
+Dear Firezone user,
+
+Here is the magic sign-in link: <%= @link %> ,
+please copy this link and open it in your browser.
```
```suggestion
Here is the magic sign-in link:

<%= @link %>

Please copy this link and open it in your browser. It is valid for 1 hour.
```

jamilbk commented on elixir/apps/web/lib/web/mailer/auth_email/sign_in_link.text.heex:
```diff
@@ -0,0 +1,9 @@
+Here is the magic sign-in link: <%= @link %> ,
+please copy this link and open it in your browser.
+
+It is valid for 1 hour.
```
```suggestion
```

jamilbk commented on elixir/apps/web/lib/web/mailer/auth_email/sign_in_link.html.heex:
```diff
@@ -0,0 +1,14 @@
+
+<p>
+  Here is the <a href={@link} target="_blank">magic sign-in link</a>. It is valid for 1 hour.
+  If you didn't request for this, you can safely discard this email.
```
```suggestion
  If you didn't request this, you can safely discard this email.
```

jamilbk committed:
Add Resources CRUD views (Supersedes #1679) (#1690)

Rebased off `andrew/live-web2`

AndrewDryga committed:
Disable dark mode

AndrewDryga committed:
Bump Vault version

jamilbk commented on elixir/apps/web/lib/web/auth.ex:
```diff
@@ -0,0 +1,192 @@
+      conn
+      |> Phoenix.Controller.put_flash(:error, "You must log in to access this page.")
+      |> maybe_store_return_to()
+      |> Phoenix.Controller.redirect(to: ~p"/#{conn.assigns.subject.account}/sign_in")
```
I think this is causing an error if `conn.assigns[:subject]` is nil

AndrewDryga committed:
Add acceptance tests for authentication

AndrewDryga committed:
Fix path crash for auth redirect

AndrewDryga committed:
Add tests for account functions

AndrewDryga committed:
Fix OpenID Connect tests

AndrewDryga committed:
Fix auth controller tests

AndrewDryga committed:
Add more tests around auth layer

AndrewDryga committed:
Fix email contents

AndrewDryga committed:
Bump all deps

AndrewDryga committed:
Start chromedriver on CI

AndrewDryga committed:
Fix trailing whitespace

AndrewDryga committed:
Rename function and format the source

AndrewDryga committed:
Check for user role for route groups

AndrewDryga committed:
Bump Elixir vsn on CI

jamilbk commented on .tool-versions:
```diff
@@ -1,8 +1,8 @@
-elixir 1.14.4-otp-25
-erlang 25.3.2
+elixir 1.15.0-otp-25
+erlang 25.3.2.2
```
Should this match `prod`?
```suggestion
erlang 26.0.1
```

AndrewDryga replied:
I built those containers to try them out but they are not used in production. We don't pull "latest tag" and our versions are fixed: https://github.com/firezone/firezone/blob/cloud/elixir/Dockerfile#L1-L3.

The reason I rolled back the Erlang version is that there is a bug in 26.0.1 that prevents elixir-ls from working (and probably other tools too), I will update CI/asdf and our containers once 26.0.2 is out.

jamilbk replied:
Ah, that's right. Thanks for clarifying.

jamilbk commented on elixir/apps/domain/lib/domain/actors/actor.ex:
```diff
@@ -4,9 +4,7 @@ defmodule Domain.Actors.Actor do
-    # TODO:
-    # field :first_name, :string
-    # field :last_name, :string
+    field :name, :string
```
Thoughts on keeping all three (`name`, `first_name`, `last_name`) so that we can save them if the `given_name` and `family_name` claims are available in the token? The use case would be the ability to sort by last name in the UI and exported logs. I suspect most orgs will have these defined.

jamilbk replied:
This PR is already quite large so maybe we can table this for a later one

AndrewDryga replied:
I would not keep them at all. Having optional fields that duplicate a required one is a source of confusion. Somebody will use them for query/searching or to render something in UI, for example, and then it will fail.

For example, they can not be used to "order by last name" because those fields are optional which means sometimes they will be empty. The search result in such a case makes no sense because we won't even use them for rendering results. Users can always sort by `name` and if the log is exported they can string split it (probably with bugs but good enough for audit cases) and do whatever they want with it.

jamilbk replied:
Hm, are you sure `name` is required and always there? The [spec mentions the `profile` scope is optional](https://openid.net/specs/openid-connect-basic-1_0.html#Scopes), and should return the following default profile claims:

`name, family_name, given_name, middle_name, nickname, preferred_username, profile, picture, website, gender, birthdate, zoneinfo, locale, and updated_at`

For example, the default Okta OIDC app mapping maps `displayName` to `name`:

<img width="995" alt="Screenshot 2023-06-27 at 8 26 25 AM" src="https://github.com/firezone/firezone/assets/167144/9055e792-50e4-4ac8-ba43-6cd84655c9d8">

However, adding a user in Okta doesn't have a field for `displayName` by default:

<img width="714" alt="Screenshot 2023-06-27 at 8 27 05 AM" src="https://github.com/firezone/firezone/assets/167144/fcb07018-4e70-418c-87bd-6c329c70aec0">

Which means `name` will be missing I believe.

jamilbk replied:
Or were you thinking we'd just concatenate `given_name` and `family_name` to populate the `name` field in our DB if needed?

AndrewDryga replied:
@jamilbk Of course, `name` is not always there but it will be available in more cases than other fields, because a lot of platforms don't ask for first/last names separately. People names are a hard problem that is very locale-specific, so I would try to avoid storing separate fields unless we do really need them. Also, syncing first/last names and joining them to a name is way easier than doing it in reverse.

Should we make it require though? It's a hard decision, from one point our UI will suffer a lot because you can't really tell who is who from user ids/group ids. We can't ask the user to enter their own name when IdP sync is enabled either, because then we will create sync issues (how to ignore certain fields) and possibly social engineering attack vectors (hey Mike can you please let me access X, it's your boss Brian).

BTW: We don't read it from claims directly but rather request from `userinfo` endpoint.

AndrewDryga replied:
> Or were you thinking we'd just concatenate given_name and family_name to populate the name field in our DB if needed?

Yes, if there is a `name` we use it, otherwise we concat.

jamilbk commented on elixir/apps/domain/lib/domain/auth/adapters/openid_connect.ex:
```diff
@@ -110,17 +114,24 @@ defmodule Domain.Auth.Adapters.OpenIDConnect do
         end
 
-      Identity.Query.by_id(identity.id)
+      # we try to use `email` if it's exposed, but fall back to OpenID Connect `sub` claim
```
This is a good call, however I wonder if we should just have a hard requirement to require the `email` claim to be there. We could always relax this requirement later on. If we can assume that each user has an email it could simplify some other features in the product that will rely on outbound emails.

There have been a couple cases where the `email` claim was missing:

- Most commonly on Azure: https://github.com/firezone/firezone/issues/1684 but the workaround/fix is fairly straightforward there
- https://github.com/firezone/firezone/issues/1572 Authgear doesn't seem to be widely used and doesn't seem to implement the spec properly here, so we can avoid building for this edge case

AndrewDryga replied:
@jamilbk I'm not following. `email` claim is optional, it will be used if it's present and fail back to `sub` otherwise (which is always present), so there is no hard requirement.

jamilbk replied:
I was suggesting perhaps we make the product decision to keep the `email` claim as a hard requirement in Firezone (and [error out](https://github.com/firezone/firezone/pull/1686/files#diff-98b6cdb69070ce6f00e9d998aea28a8889d2a38434b9aff2c653fc8e95a8799fR41) if it's missing) because it will make a number of features/edge cases easier to reason about:

- We can assume every user can receive an outbound emails for things like notifications, reports
- Admin signs up with magic link, connects an IdP, then migrates his/her account to a connected IdP
- Admin A signs up, makes user B an Admin, then user B demotes Admin A. Do we allow user B to be the sole admin of this tenant with no email?
- Billing reminders, invoices, etc

The two examples I gave in my previous comment were cases where our current hard requirement for `email` may have caused an issue, but in the end were resolved without issue.

We could also prompt for an email each time we encounter the scenarios above, but that might get messy?

AndrewDryga replied:
Ah, now I get it. 

I'm not a big fan of using email claims at all, orgs have a policy for email reuse but even then there is a grace period, (eg. you will be able to reuse it in 2 years), while in Firezone we don't know about it and will eventually log a user in into somebody else account. It's even worse if the company has no email reuse policy.

So maybe instead we can drop `email` claim entirely, use `sub` which is part of the spec, unique, and always there, and just use `email` to fill out the profile for notifications (and warn if there is none to manually enter it)?

jamilbk replied:
Email reuse policies are an interesting edge case I hadn't thought of.

Yeah, we can stick to the spec as best we can, and then require emails when needed if they're missing. Another edge case I thought was for support -- we'll need an email to communicate with the user if he/she opens a support ticket. 🙃 

In cases like this, we can require the Admin to add the user's email to the tenant in Azure for example so we pick it up on the next SCIM sync.

bmanifold replied:
@AndrewDryga, I agree with your thoughts on email reuse and we should use the `sub` claim for identification on the Firezone side of things.  That said, I can also understand where @jamilbk is coming from since we may need to be able to communicate with these users outside of the Firezone application.

I may be thinking about this wrong, but I think we could do both.  We could use the `sub` claim as the Firezone identifier and also require that the `email` claim be populated too if we wanted.  We shouldn't need to worry about logging someone in to an old account at that point, because the IDP would only ever give the `sub` id for the currently active owner of the email right?

If what I've outlined above holds true, businesses may like the ability to control what email is being set for any given Firezone account they "own", rather than allowing the user to fill in whatever email they want in their profile.  e.g. If I'm an employee of `Company A`, I probably shouldn't be allowed to set my Firezone profile email to `user@gmail.com`, but rather be forced to use my company email `user@companya.com`

I'm sure there's probably something I'm not thinking through, but figured I'd throw this out and see what everyone thought

jamilbk replied:
That's a good point @bmanifold. I think for the purposes of this PR we should probably clarify that OIDC claims are only used for sign in, so it does make sense to use `sub` to find the user and match on it for that.

The tricky part is the SCIM integration. For sync, _if_ we find a user's email matching a particular `sub` we can populate it, but if not, we can't really do anything at that point. We probably still want to allow the user to authenticate via OIDC, but we're at the mercy of the sync'd attributes from SCIM to populate the email.

I think @AndrewDryga's suggestion to allow them to populate the email was good, but perhaps we require the admin to set this up successfully via the SCIM connector as instead of letting the user do it as @bmanifold suggested.

For reference, I did [some research](https://github.com/firezone/product/issues/461) on which providers support SCIM over OIDC and which don't (or require SAML):

| Provider | SCIM support |
| --- | --- |
| Auth0 | Supported by [professional services only](https://auth0.com/docs/authenticate/protocols/scim). GA expected Q4 '23 |
| Google | Only supported for [apps in the directory](https://apps.google.com/supportwidget/articlehome?hl=en&article_url=https%3A%2F%2Fsupport.google.com%2Fa%2Fanswer%2F7681608%3Fhl%3Den&assistant_id=generic-unu&product_context=7681608&product_name=UnuFlow&trigger_context=a). Needs a custom adapter otherwise |
| Okta | [SAML only](https://help.okta.com/en-us/Content/Topics/Apps/Apps_App_Integration_Wizard_SCIM.htm) |
| Azure AD | Supported for [both OIDC and SAML](https://learn.microsoft.com/en-us/azure/active-directory/app-provisioning/use-scim-to-provision-users-and-groups) |
| JumpCloud | [Supported for SAML](https://jumpcloud.com/support/provision-and-manage-users-and-groups-in-apps-using-custom-scim-identity-management-integration) |
| OneLogin | [Supported for SAML](https://onelogin.service-now.com/support?id=kb_article&sys_id=1c8d87f2dbcfad1057fac24505961922) |
| AWS IAM Identity Center | [Supported for SAML](https://docs.aws.amazon.com/singlesignon/latest/userguide/scim-profile-saml.html) |
| Authentik | Supported for [both OIDC and SAML](https://goauthentik.io/docs/providers/scim/) |
| Keycloak | Not supported |
| Ory Stack | [Not supported](https://github.com/ory/kratos/issues/1028) |

jamilbk commented on elixir/apps/domain/lib/domain/auth/identity/query.ex:
```diff
@@ -4,6 +4,9 @@ defmodule Domain.Auth.Identity.Query do
     |> where([identities: identities], is_nil(identities.deleted_at))
+    |> join(:inner, [identities: identities], actors in assoc(identities, :actor), as: :actors)
+    |> where([actors: actors], is_nil(actors.deleted_at))
+    |> where([actors: actors], is_nil(actors.disabled_at))
```
👍 Disabling a user is relatively common in current Firezone, thanks for keeping this around. The most common use case I've seen this pop up is when an admin wants to only temporarily disable access for a particular user.

jamilbk commented on elixir/apps/domain/lib/domain/validator.ex:
```diff
@@ -21,7 +21,9 @@ defmodule Domain.Validator do
   def validate_email(changeset, field) do
-    validate_format(changeset, field, ~r/@/, message: "is invalid email address")
+    changeset
+    |> validate_format(field, ~r/^[^\s]+@[^\s]+$/, message: "is invalid email address")
```
```suggestion
    |> validate_format(field, ~r/^[^\s]+@[^\s]+$/, message: "is an invalid email address")
```

jamilbk commented on elixir/apps/domain/test/domain/auth/adapters/userpass_test.exs:
```diff
@@ -50,7 +50,7 @@ defmodule Domain.Auth.Adapters.UserPassTest do
       assert errors_on(changeset) == %{
                provider_virtual_state: %{
-                 password: ["should be at least 12 character(s)"],
+                 password: ["should be at least 12 byte(s)"],
```
So curious regarding this change 😆

AndrewDryga replied:
@jamilbk Some hashing functions (like bcrypt) are having a hard limit in bytes, not characters (UTF can span multiple bytes). We use Argon which has a limit of 4096 bytes, but I decided to keep a low realistic limit in there and make sure it's calculated properly for the future so that we don't get confused by it later.

AndrewDryga replied:
And I did not bother about UI complications because I believe we don't plan to use userpass in production.

jamilbk commented on elixir/apps/web/lib/web/auth.ex:
```diff
@@ -0,0 +1,221 @@
+    {:ok, session_token} = Domain.Auth.create_session_token_from_subject(subject)
+
+    conn
+    |> Plug.Conn.put_session(:logged_in_at, DateTime.utc_now())
```
(nit) 
```suggestion
    |> Plug.Conn.put_session(:signed_in_at, DateTime.utc_now())
```

jamilbk commented on elixir/apps/web/lib/web/auth.ex:
Lots of critical paths in this module -- thanks for documenting it so well and making it easy to follow.

jamilbk commented on elixir/apps/web/lib/web/components/core_components.ex:
Just leaving a note here for @bmanifold -- probably will be getting intimately familiar with this module as you work on UI. Many of these were generated by Phoenix by default, so most of them will need a little tweaking to fit our Flowbite-powered design system, and some aren't used at all (yet).

bmanifold replied:
@jamilbk thanks for the heads up

jamilbk commented on elixir/apps/web/test/support/acceptance_case/auth.ex:
```diff
@@ -32,70 +32,82 @@ defmodule Web.AcceptanceCase.Auth do
+      cookie =
+        %{
+          "session_token" => token,
+          "logged_in_at" => DateTime.utc_now(),
```
(nit)

```suggestion
          "signed_in_at" => DateTime.utc_now(),
```

jamilbk commented on elixir/apps/web/lib/web/controllers/auth_controller.ex:
```diff
@@ -0,0 +1,217 @@
+      redirect_to = get_session(conn, :user_return_to) || Auth.signed_in_path(subject)
+
+      conn
+      |> Web.Auth.renew_session()
```
👍

jamilbk commented on elixir/apps/web/lib/web/controllers/auth_controller.ex:
Thanks again for documenting this so well

jamilbk commented on elixir/apps/web/lib/web/mailer/auth_email/sign_in_link.html.heex:
```diff
@@ -0,0 +1,14 @@
+</p>
+
+<p>
+  Here is the <a href={@link} target="_blank">magic sign-in link</a>. It is valid for 1 hour.
```
(nit)

```suggestion
  Here is the <a href={@link} target="_blank">magic sign-in link</a> you requested. It is valid for 1 hour.
```

jamilbk commented on elixir/apps/web/lib/web/mailer/auth_email/sign_in_link.text.heex:
```diff
@@ -0,0 +1,8 @@
+
+Dear Firezone user,
+
+Here is the magic sign-in link:
```
(nit) 
```suggestion
Here is the magic sign-in link you requested:
```

jamilbk reviewed (commented):
Looks great overall. Left a few comments/suggestions/questions.

Just waiting on CI to turn green 😁

AndrewDryga committed:
Address code review comments

AndrewDryga committed:
Remove ssl_verify_fun override

It was needed due to compatibility issues with OTP 26 but maintainer already pushed a fix

AndrewDryga committed:
Recompile cache when elixir/otp version changed

AndrewDryga committed:
Fix compile warnings

AndrewDryga committed:
Fix more warnings

AndrewDryga committed:
Rename test step

AndrewDryga committed:
Do not use email claim as user identifier

AndrewDryga committed:
Fix Elixir 1.15 test compilation issues

AndrewDryga committed:
Remove unused deps

AndrewDryga committed:
Fix acceptance tests
